cdisc_rules_engine/services/data_services/usdm_data_service.py

```diff
@@ -0,0 +1,310 @@
+        datasets = self.__get_datasets(json)
+        dataset_paths = [
+            path
+            for dataset_metadata in datasets
+            if dataset_metadata["domain"] == dataset_name
+            for path in dataset_metadata["full_path"]
+        ]
+        all_nodes = []
+        for dataset_path in dataset_paths:
+            nodes = [match for match in parse(dataset_path["path"]).find(json)]
+            if len(nodes) != 1:
+                raise Exception(
+                    f"Multiple objects found with path: {dataset_path['path']}"
+                )
+            node = nodes[0]
+            if dataset_path["type"] == "reference":
+                node.value = self.__find_definition(json, node.value)
+                node.type = "reference"
+            else:
+                node.type = "definition"
+            all_nodes.append(node)
+        records = [
+            self.__get_record_metadata(node) | self.__get_record_data(node.value)
+            for node in all_nodes
+        ]
+        return pandas.DataFrame(records)
+
+    def __find_definition(self, json, id: str):
+        definition = parse(f"$..*[?(@.id = '{id}')]").find(json)
+        if definition:
+            if len(definition) > 1:
+                raise Exception(f"Multiple objects found with id: {id}")
+            return definition[0].value
+        return None
+
+    def __get_entity_name(self, value, path: This):
+        if type(value) is dict:
+            api_type = (
+                value.get("instanceType") if "instanceType" in value else f"{path}"
+            )
+        else:
+            # primitive types
+            api_type = value.__class__.__name__
+        return self.entity_dict.get(api_type, api_type)
+
+    def __read_metadata(
+        self,
+        json,
+        parent_node: DatumInContext,
+        child_value,
+        full_path: str,
+    ):
+        ty = "definition"
+        if type(child_value) is str and (
+            f"{parent_node.path}".endswith("Id")
+            or f"{parent_node.path}".endswith("Ids")
+        ):
+            definition = self.__find_definition(json, child_value)
+            if definition:
+                child_value = definition
+                ty = "reference"
+        entity_name = self.__get_entity_name(child_value, parent_node.path)
+        return {
+            "path": full_path,
+            "entity": entity_name,
+            "type": ty,
+        }
+
+    def __get_datasets(self, json) -> List[dict]:
+        """
+        This is a bit convoluted because there is a bug in jsonpath_ng
+        where this query does not return object values within an array
+        """
+        metadata = []
+        for node in parse("$..*").find(json):
+            if type(node.value) is list:
+                for index, child in enumerate(node.value):
+                    if metadatum := self.__read_metadata(
+                        json, node, child, f"{node.full_path}[{index}]"
```

Comment:
I updated this for consistency, but the opposite way. Removed the extra periods that jsonpath-ng creates so that we are using the more common (and correct?) representation.

Reply:
I was also assuming that not having a dot before indexes was the "correct" representation - that seems to be what the JSONPath documentation says.  However, the dots are included in the jsonpath-ng documentation, so I guess they'd be considered correct if you're using jsonpath-ng.  The changes look good.